### PR TITLE
Issue#13 Implements LIC4

### LINKED ISSUE #13: Fix LIC 4
This feature will evaluate if the following is true or false for the input data points:

There exists at least one set of Q PTS consecutive data points that lie in more than QUADS
quadrants. Where there is ambiguity as to which quadrant contains a given point, the priority
of decision will be by quadrant number, i.e., I, II, III, IV. For example, the data point (0,0)
is in quadrant I, the point (-l,0) is in quadrant II, the point (0,-l) is in quadrant III, the point
(0,1) is in quadrant I and the point (1,0) is in quadrant I.

(2 <= Q_PTS <= NUMPOINTS), (1 <= QUADS <= 3)

### PR DESCRIPTION
**Pull request by joakimAbdinurIusuf**
Implements LIC4() and several helper methods to check if LIC4 is true or false. Also implements 8 unit tests; 4 check positive and negative cases and the other four check that LIC4 is given valid input.

closes #13 #51

### CODE REVIEW

**Review by asta12 — COMMENTED** [review]
Is IllegalArgumentException a good choice? The function takes no arguments. (This comment was meant to be connected to some code, I made another one)

File: src/main/java/Decide.java [hunk 1]
```diff
@@ -79,6 +79,22 @@ public boolean LIC3() throws IllegalArgumentException {
...
 
+    /**
+     * Check if LIC4 is true.
+     *
+     * @return true if there exists at least one set of Q_PTS consecutive data points
+     * that lie in more than QUADS quadrants, false otherwise.
+     */
+    public boolean LIC4() throws IllegalArgumentException {
+        if (parameters.getQ_PTS() < 2) {
+            throw new IllegalArgumentException("Q_PTS must be greater than or equal to 2.");
```
Comment (asta12) [inline src/main/java/Decide.java]: New try, I forgot to comment in the code. I am still learning :). Is IllegalArgumentException a good choice? The function takes no arguments.

File: src/test/java/DecideTest.java [hunk 2]
```diff
@@ -80,6 +87,90 @@ public void givenAREA1IsLessThanZero_whenLIC3_thenExceptionIsThrownAndMessagePri
...
+    @Test
+    public void givenExistsQ_PTSConsecutivePointsThatLieInMoreThanQUADSQuadrantsClear_whenLIC4_thenAssertTrue() {
+        double[][] points = new double[][]{{1.0, 1.0}, {-1.0, 1.0}, {-1.0, -1.0}};
+        int numpoints = points.length;
+        Parameters parameters = new Parameters(0, 0, 0, 0, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+        Decide decide = new Decide(numpoints, points, parameters, (LCM)null, (boolean[])null);
+        boolean LIC5True = decide.LIC5();
+        assertTrue(LIC5True);
+    }
+    */
```
Comment (asta12) [inline src/test/java/DecideTest.java]: Commented out code should not be pushed.

**Review by asta12 — COMMENTED** [review]
There are multiple places where IllegalArgumentException is used.

File: src/test/java/DecideTest.java (hunk 2 above)
Comment (asta12) [reply src/test/java/DecideTest.java]: I saw that this was removed in a later commit. Resolving this.

File: src/main/java/Decide.java [hunk 3]
```diff
@@ -81,6 +81,75 @@ public boolean LIC3() throws IllegalArgumentException {
...
+    }
+
+    // If (x>0, y<0) then the data point is in quadrant 4.
+    private boolean dataPointIsInFourthQuadrant(int i, int j) {
+        return points[i + j][0] > 0 && points[i + j][1] < 0;
+    }
+
+    // If (0, y<0) or (x<0, y<0) then the data point is in quadrant 3.
+    private boolean dataPointIsInThirdQuadrant(int i, int j) {
+        return (points[i + j][0] == 0 && points[i + j][1] < 0) || (points[i + j][0] < 0 && points[i + j][1] < 0);
```
Comment (asta12) [inline src/main/java/Decide.java]: Why not have `points[i + j][0] <= 0 && points[i + j][1] < 0` ? Is there any specific reason as to why you have made two cases and use OR between them? Same goes for the other quadrants as well.

File: src/main/java/Decide.java (hunk 1 above)
Comment (joakimAbdinurIusuf) [reply src/main/java/Decide.java]: No, since the method doesn't take any arguments. Created a custom exception called IllegalParameterObjectException and changed all the places where IllegalArgumentException was used to use this new exception.

File: src/main/java/Decide.java (hunk 3 above)
Comment (joakimAbdinurIusuf) [reply src/main/java/Decide.java]: I wrote down the cases in which different values were in a certain quadrant and coded them. Had not thought of this improvement, thanks!

**Comment by asta12** [comment]
Looks good! I will merge
